src/lsp.h

```diff
@@ -0,0 +1,59 @@
+/*
+ *      lsp.h - this file is part of Geany, a fast and lightweight IDE
+ *
+ *      Copyright 2023 The Geany contributors
+ *
+ *      This program is free software; you can redistribute it and/or modify
+ *      it under the terms of the GNU General Public License as published by
+ *      the Free Software Foundation; either version 2 of the License, or
+ *      (at your option) any later version.
+ *
+ *      This program is distributed in the hope that it will be useful,
+ *      but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *      GNU General Public License for more details.
+ *
+ *      You should have received a copy of the GNU General Public License along
+ *      with this program; if not, write to the Free Software Foundation, Inc.,
+ *      51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef GEANY_LSP_H
+#define GEANY_LSP_H 1
+
+#include "document.h"
+
+G_BEGIN_DECLS
+
+typedef struct {
+	gboolean (*autocomplete_available)(GeanyDocument *doc);
+	void (*autocomplete_perform)(GeanyDocument *doc);
+
+	gboolean (*calltips_available)(GeanyDocument *doc);
+	void (*calltips_show)(GeanyDocument *doc);
+
+	gboolean (*goto_available)(GeanyDocument *doc);
+	void (*goto_perform)(GeanyDocument *doc, gboolean definition);
+} Lsp;
+
+
+void lsp_register(Lsp *lsp);
+void lsp_unregister(Lsp *lsp);
```

Comment:
I would also pass the size of the `Lsp`. Reason: If you extend the struct for new function points you can't be sure if the plugin that's registering is already compiled against the new size, unless you also bump the ABI. To avoid having to bump the ABI the plugin can pass `sizeof(Lsp)` that it was compiled against, and Geany can pad with NULLs up to the actual size.

Furthermore, these kind of register functions should also pass the `GeanyPlugin` to give Geany some context.

Reply:
> I would also pass the size of the Lsp

Oh, I somehow didn't post the right version of the patch - in the correct one I also added `gchar padding[1024]` at the end of the struct. Plugins will have to allocate this struct with `g_new0()` or on the heap so the rest of the struct is zero and those `CALL_IF_EXISTS()` macros check the NULL and fall back to some empty implementation.

> Furthermore, these kind of register functions should also pass the GeanyPlugin to give Geany some context.

Could you add more detail what this would be good for?